### ISSUE TITLE
[Submission] {6} params - {zcbtrak} trained

### ISSUE THREAD

**Issue by zcbtrak** [anadim/AdderBoard#76]
### Author

Jarek Rakoto-Miklas (@zcbtrak)

### Unique Parameter Count

6

### Accuracy

100

### Method

Trained (Coordinate Descent)

### Architecture

1L Qwen-derived decoder, d=2, 1h, hd=2, ff=2 (fixed Q, folded norm)

### Link to Code

https://github.com/zcbtrak/6P-coordinate-descent

### Verification Output

```shell
Model: gpu_trained_6p
Author: zcbtrak
Parameters (unique): 6
Architecture: 1L Qwen-derived decoder, d=2, 1h, hd=2, ff=2 (fixed Q, folded norm)
Tricks: RoPE period-19 geometry, hardcoded Q_proj (routing as architectural constant, 0 params), norm weights folded into tied output head (0 extra params), tied carry hinge gate, shared carry-scale scalar, 2-parameter embedding e(d)=[c0-c1*d^2,-d], gradient-free coordinate descent training, landscape-aware parameter perturbations

  Progress: 1000/10010 (1000/1000 correct) [7.3s]
  Progress: 2000/10010 (2000/2000 correct) [13.7s]
  Progress: 3000/10010 (3000/3000 correct) [20.4s]
  Progress: 4000/10010 (4000/4000 correct) [26.9s]
  Progress: 5000/10010 (5000/5000 correct) [33.5s]
  Progress: 6000/10010 (6000/6000 correct) [40.5s]
  Progress: 7000/10010 (7000/7000 correct) [47.7s]
  Progress: 8000/10010 (8000/8000 correct) [54.4s]
  Progress: 9000/10010 (9000/9000 correct) [61.2s]
  Progress: 10000/10010 (10000/10000 correct) [67.7s]

Results: 10010/10010 correct (100.00%)
Time: 67.8s (148 additions/sec)
Status: QUALIFIED (threshold: 99%)
```

## Architecture

Identical architecture to the 6-parameter hand-coded submission (same forward
pass, same fixed constants), but the 6 weights were discovered from scratch by
a generic GPU-accelerated coordinate descent optimizer — no analytical
derivation, no warm-starting from the hand-coded solution.

### The 6 learned parameters

| # | Name | Trained value | Hand-coded value | Role |
|---|---|---|---|---|
| 1 | embed_w0 | 524.279 | 1000.0 | Embedding bias |
| 2 | embed_w1 | 0.001588 | 0.001 | Embedding quadratic |
| 3 | v_proj_w | −6689.18 | −15556.3 | V projection scalar |
| 4 | gate_w0 | −17615.8 | −12032.0 | Tied gate threshold |
| 5 | gate_w1 | 124018.0 | 128000.0 | Tied gate slope |
| 6 | carry_w | 0.3374 | 0.3906 | Carry projection scalar |

The optimizer found a **completely different point in the loss landscape** from
the hand-coded solution (embed_w0 differs by ~2x, v_proj_w by ~2.3x), which
demonstrates the 6-parameter architecture has multiple basins of attraction and
the solution is genuinely learnable by a generic algorithm.

### Training algorithm

GPU coordinate descent with random restarts, implemented in PyTorch:
- Cyclic 1D grid search over each parameter
- Range narrowing with shrink factor
- Random restarts from broad parameter ranges
- No curriculum, no warm-starting from analytical solution
- The optimizer sees the same `forward()` as verify.py (autoregressive generation)

This is a fully generic optimizer — it works with any differentiable-or-not
model and does not encode any knowledge of addition.

### Parameter counting justification

Same architecture as the hand-coded 6p entry. The 0-param constants are:
- **Q_proj** `[cos(Φ), −sin(Φ)]`: RoPE-derived routing angle (PE exemption)
- **Norm** `[50√2, −10√2]`: folded into tied output head
  (`rms_norm(h,w)@E^T ≡ unit_rms_norm(h)@(w⊙E)^T`)
- **K_proj** `[x₀, 0]`, **O_proj** `[0, x₀]`: sparse projections
- **Gate tie coefficient** `0.001`: architectural constant in gate formula

These constants were **fixed throughout training** — the optimizer never
searched over them. The fact that it found 100%-accurate parameters with all
constants frozen confirms they are architectural, not hidden parameters.

See the companion hand-coded submission for full precedent analysis.

### Key tricks

- RoPE period-19 geometry
- Hardcoded Q_proj (0 params)
- Folded norm (0 extra params)
- Tied carry hinge gate
- Shared carry-scale scalar
- 2-parameter embedding `e(d) = [c₀ − c₁·d², −d]`
- GPU coordinate descent (generic optimizer, no domain knowledge in search)
